system/security/shared/caching.cpp

```diff
@@ -742,16 +757,22 @@ void CPermissionsCache::flush()
             delete (*ui).second;
         m_userCache.clear();
     }
+
     if (m_useLegacyDefaultFileScopePermissionCache)
     {
+        // MORE - is this safe? m_defaultPermission is unprotected,
+        // and m_defaultPermission could be changed while cleared here - does that matter?
         m_defaultPermission = SecAccess_Unknown;
     }
     else
     {
         CriticalBlock defaultScopePermissionBlock(syncDefaultScopePermissions);
         m_userDefaultFileScopePermissions.clear();
     }
-    m_lastManagedFileScopesRefresh = 0;
+
+    // If the managed file scope cache fill thread is running, signal it to refill the cache
+    if (m_fileScopeCacheFillThread.joinable())
```

Comment:
This is a semantic change.
Previously a call to flush(), immediately followed by a call to queryPermsManagedFileScope() would guarantee new refreshed results. Now it does not.

As it stands, flush() immediately followed by queryPermsManagedFileScope() would likely return the old stale results.

Options:
1) Ignore semantic differences, results will be refreshed 'soon'.
2) Mark a flag, that is reset when done, but causes queryPermsManagedFileScope() to block if flush refresh is pending.

Reply:
I lean towards option 1. The reason is we are trying to avoid threads waiting for a flush to finish. I considered this scenario when adding the refresh thread.

Reply:
I don't think that's right, and introduces a new potential source of subtle future confusion.

If clearPermissionsCache() is called and then a file scope permission is looked up in the time frame that it's refreshing (which we know can take some time on big systems), then the wrong result is returned..

It's not just that a stale result is returned, it's the wrong result (default user perms).